content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

```diff
@@ -751,26 +751,154 @@ with the response:
-- Explain + and == for list maps and list sets (table? whatever looks better)
-- DO NOT: provide all the motivation and design rationale from the KEP.
+* spec: Invalid value: map[string]interface {}{"maxReplicas":10, "minReplicas":0, "replicas":20}: failed rule: self.replicas <= self.maxReplicas
+```
+
+`x-kubernetes-validations` could have multiple rules. 
+
+The `rule` under `x-kubernetes-validations` represents the expression which will be evaluated by CEL.
+
+The `message` represents the message displayed when validation fails. The message is required if the Rule contains
+line breaks. The message must not contain line breaks. If unset, the message is "failed rule: {Rule}".
+
+Validation rules are compiled when CRDs are created/updated. 
+Compilation error will be returned even feature gate is disabled.
+
+Example:
+
+| Rule                                     |      Message                                                                                                                                                                         |
+| -------------                            | ------------                                                                                                                                                                         |
+|"self == true" against integer type       | Invalid value: apiextensions.ValidationRule{Rule:"self == true", Message:""}: compilation failed: ERROR: \<input>:1:6: found no matching overload for '_==_' applied to '(int, bool)'|
+|"self.nonexistingValue > 0"               | Invalid value: apiextensions.ValidationRule{Rule:"self.nonexistingValue > 0", Message:""}: compilation failed: ERROR: \<input>:1:5: undefined field 'nonexistingValue'                |
+
+The Rule is scoped to the location of the `x-kubernetes-validations` extension in the schema.
+
+The `self` variable in the CEL expression is bound to the scoped value.
+
+Example Validation Rules:
+
+| Rule                                                                             | Purpose                                                                           |
+| ----------------                                                                 | ------------                                                                      |
+| `self.minReplicas <= self.replicas && self.replicas <= self.maxReplicas`         | Validate that the three fields defining replicas are ordered appropriately        |
+| `'Available' in self.stateCounts`                                                | Validate that an entry with the 'Available' key exists in a map                   |
+| `(size(self.list1) == 0) != (size(self.list2) == 0)`                             | Validate that one of two lists is non-empty, but not both                         |
+| `!('MY_KEY' in self.map1) || self['MY_KEY].matches('^[a-zA-Z]*$')`               | Validate the value of a map for a specific key, if it is in the map               |
```

Comment:
Looks like I missed the closing `'` in `self['MY_KEY]` in the KEP, let's fix it here.

Reply:
Also pipe(|) needs to be escaped so it would display properly. Updated.